File: scripts/ci_state_guard.mjs
#!/usr/bin/env node
/**
 * CI drift / arch guard — validate the frontend ↔ /state contract.
 *
 * Parses src/lib/stores/telemetry.ts and enforces:
 *   1. The TelemetryData interface has NOT lost any fields it currently has
 *      (prevents accidental regressions).
 *   2. The applyLiveFrame function still reads all previously-consumed fields.
 *   3. Backend fields not yet consumed by the frontend are WARNINGS only.
 *
 * The backend's 14-field contract is guarded separately in the helios repo.
 * This guard protects the FRONTEND from accidentally dropping fields it
 * already consumes.
 *
 * Usage: node scripts/ci_state_guard.mjs
 */

import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const TELEMETRY_TS = resolve(
  import.meta.dirname || fileURLToPath(new URL('.', import.meta.url)),
  '../src/lib/stores/telemetry.ts'
);

// Fields currently in TelemetryData.  If any of these are removed in a PR,
// this guard FAILS.  New fields may be added freely.
export const EXPECTED_IFACE_FIELDS = [
  'uptime_seconds',
  'tick_rate',
  'pam',
  'rpd_used',
  'rpd_budget',
  'provider',
  'connected',
  'is_speaking',
  'neurobus',
  'neurobusHistory',
  'hidden_state',
  'runtime'
];

// Fields that applyLiveFrame() currently reads from the frame.  If any are
// removed, this guard FAILS.
export const EXPECTED_LIVE_FRAME_FIELDS = [
  'neurobus',
  'uptime_seconds',
  'tick_rate',
  'pam',
  'rpd_used',
  'rpd_budget',
  'provider',
  'is_speaking',
  'hidden_state',
  'runtime'
];

/**
 * @param {string} source
 * @returns {string[] | null}
 */
export function extractInterfaceFields(source) {
  const ifaceMatch = source.match(/interface\s+TelemetryData\s*\{([\s\S]*?)\n\s*\}/);
  if (!ifaceMatch) return null;

  const body = ifaceMatch[1];
  const fields = [];
  const fieldRegex = /^\s+(\w+)\s*[?:]/gm;
  let m;
  while ((m = fieldRegex.exec(body)) !== null) {
    fields.push(m[1]);
  }
  return fields;
}

/**
 * @param {string} source
 * @returns {string[] | null}
 */
export function extractApplyLiveFrameFields(source) {
  const fnMatch = source.match(
    /const\s+applyLiveFrame\s*=\s*\([^)]*\)\s*=>\s*\{([\s\S]*?)\n\s*\};/
  );
  if (!fnMatch) return null;

  const body = fnMatch[1];
  const fields = [];
  const frameFieldRegex = /frame\.(\w+)/g;
  let m;
  const seen = new Set();
  while ((m = frameFieldRegex.exec(body)) !== null) {
    if (!seen.has(m[1])) {
      seen.add(m[1]);
      fields.push(m[1]);
    }
  }
  return fields;
}

/**
 * @param {string} source
 * @returns {{ ifaceFields: string[], liveFrameFields: string[], errors: string[] }}
 */
export function validateContract(source) {
  const errors = [];
  const ifaceFields = extractInterfaceFields(source);
  if (!ifaceFields) {
    throw new Error('ERROR: could not extract TelemetryData interface');
  }

  const liveFrameFields = extractApplyLiveFrameFields(source);
  if (!liveFrameFields || liveFrameFields.length === 0) {
    throw new Error('ERROR: drift guard could not locate the live-frame fields — refusing to pass');
  }

  // Check TelemetryData hasn't lost any fields
  for (const field of EXPECTED_IFACE_FIELDS) {
    if (!ifaceFields.includes(field)) {
      errors.push(`TelemetryData missing field "${field}" (was present before)`);
    }
  }

  // Check applyLiveFrame hasn't lost any field reads
  for (const field of EXPECTED_LIVE_FRAME_FIELDS) {
    if (!liveFrameFields.includes(field)) {
      errors.push(`applyLiveFrame no longer reads "${field}" (was consumed before)`);
    }
  }

  return { ifaceFields, liveFrameFields, errors };
}

function main() {
  let source;
  try {
    source = readFileSync(TELEMETRY_TS, 'utf-8');
  } catch {
    console.error(`ERROR: cannot read ${TELEMETRY_TS}`);
    process.exit(1);
  }

  try {
    const { ifaceFields, liveFrameFields, errors } = validateContract(source);
    console.log(`TelemetryData fields: ${ifaceFields.join(', ')}`);
    console.log(`applyLiveFrame reads:  ${liveFrameFields.join(', ')}`);

    if (errors.length > 0) {
      console.error('\n*** DRIFT GUARD FAILED ***');
      for (const e of errors) {
        console.error(`  - ${e}`);
      }
      console.error(
        '\nTo fix: if the field removal is intentional, update\n' +
          'EXPECTED_IFACE_FIELDS / EXPECTED_LIVE_FRAME_FIELDS in this script\n' +
          "and get Navpreet's approval."
      );
      process.exit(1);
    }
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    console.error(message);
    process.exit(1);
  }

  console.log('\nOK — frontend ↔ /state contract is intact (no regressions).');
}

const isMain =
  process.argv[1] && resolve(fileURLToPath(import.meta.url)) === resolve(process.argv[1]);
if (isMain) {
  main();
}

File: scripts/ci_state_guard.py
#!/usr/bin/env python3
"""CI drift / arch guard — enforce the /state contract.

Parses aegis/web/server.py and extracts the keys returned by _build_state().
Fails if:
  - Any frozen field has been removed (breaking change)

Warns (but passes) if:
  - New fields are added (additive — needs review but not a CI blocker)

Usage: python scripts/ci_state_guard.py
"""
from __future__ import annotations

import ast
import re
import sys
from pathlib import Path

SERVER_PY = Path(__file__).resolve().parent.parent / "aegis" / "web" / "server.py"

# The frozen set of top-level keys that _build_state() MUST return.
EXPECTED_FIELDS: frozenset[str] = frozenset({
    "neurobus",
    "h",
    "is_speaking",
    "last_action_type",
    "tick_id",
    "mnemosyne_event",
    "provider",
    "rpd_used",
    "rpd_budget",
    "connected",
    "uptime_seconds",
    "tick_rate",
    "pam",
    "coherence",
    "runtime",
})


def _extract_return_keys_from_source(source: str) -> list[str] | None:
    """Find the return dict literal inside _build_state() and extract its keys.

    Uses AST parsing to avoid fragile regex.  Returns None if the function
    or return statement can't be found.
    """
    tree = ast.parse(source)
    for node in ast.walk(tree):
        if not isinstance(node, ast.FunctionDef) or node.name != "_build_state":
            continue
        # Walk the function body for a Return statement with a Dict value.
        for stmt in ast.walk(node):
            if not isinstance(stmt, ast.Return) or not isinstance(stmt.value, ast.Dict):
                continue
            keys: list[str] = []
            for key_node in stmt.value.keys:
                if isinstance(key_node, ast.Constant) and isinstance(key_node.value, str):
                    keys.append(key_node.value)
                else:
                    # Non-constant key (computed) — abort
                    return None
            return keys
    return None


def _fallback_regex_extract(source: str) -> list[str] | None:
    """Fallback: regex-extract string keys from the return dict in _build_state().

    Only used if AST parsing fails (e.g. the source has syntax the parser
    doesn't understand).
    """
    # Find _build_state function body, then the return dict
    pattern = re.compile(
        r'def _build_state\b.*?return\s*\{(.*?)\}',
        re.DOTALL,
    )
    m = pattern.search(source)
    if not m:
        return None
    body = m.group(1)
    # Extract string literal keys: "key": ...
    keys = re.findall(r'"(\w+)"\s*:', body)
    return keys if keys else None


def main() -> int:
    if not SERVER_PY.exists():
        print(f"ERROR: {SERVER_PY} not found", file=sys.stderr)
        return 1

    source = SERVER_PY.read_text(encoding="utf-8")
    keys = _extract_return_keys_from_source(source)
    if keys is None:
        keys = _fallback_regex_extract(source)
    if keys is None:
        print("ERROR: could not extract _build_state() return keys", file=sys.stderr)
        return 1

    actual = frozenset(keys)

    print(f"Detected {len(keys)} /state fields: {sorted(keys)}")

    errors: list[str] = []
    warnings: list[str] = []

    removed = EXPECTED_FIELDS - actual
    if removed:
        errors.append(f"Fields REMOVED (breaking): {sorted(removed)}")

    added = actual - EXPECTED_FIELDS
    if added:
        warnings.append(f"Fields ADDED (additive, needs review): {sorted(added)}")

    if warnings:
        print("\n*** DRIFT GUARD WARNINGS ***")
        for w in warnings:
            print(f"  - {w}")

    if errors:
        print("\n*** DRIFT GUARD FAILED ***", file=sys.stderr)
        for e in errors:
            print(f"  - {e}", file=sys.stderr)
        print(
            "\nTo fix: if the /state contract change is intentional, update "
            "EXPECTED_FIELDS in this script and tests/test_state_contract.py, "
            "then get Navpreet's approval.",
            file=sys.stderr,
        )
        return 1

    print("OK — /state contract is intact (frozen fields present, no removals).")
    return 0


if __name__ == "__main__":
    sys.exit(main())
